source/unified-test-format/unified-test-format.rst

```diff
@@ -0,0 +1,1896 @@
+
+.. _entity_session:
+
+- ``session``: Optional document. Corresponds with an explicit ClientSession
+  object.
+
+  The structure of this document is as follows:
+
+  - ``id``: Required string. Unique name for this entity. The YAML file SHOULD
+    define a `node anchor`_ for this field (e.g. ``id: &session0 session0``).
+
+  - ``client``: Required string. Client entity from which this session will be
+    created. The YAML file SHOULD use an `alias node`_ for a client entity's
+    ``id`` field (e.g. ``client: *client0``).
+
+  - ``sessionOptions``: Optional document. Map of parameters to pass to
+    `MongoClient.startSession <../source/sessions/driver-sessions.rst#startsession>`__
+    when creating the session. Supported options are defined in the following
+    specifications:
+
+    - `Causal Consistency <../causal-consistency/causal-consistency.rst#sessionoptions-changes>`__
+    - `Transactions <../transactions/transactions.rst#sessionoptions-changes>`__
+
+    When specifying TransactionOptions for ``defaultTransactionOptions``, the
+    transaction options MUST remain nested under ``defaultTransactionOptions``
+    and MUST NOT be flattened into ``sessionOptions``.
+
+- ``bucket``: Optional document. Corresponds with a GridFS Bucket object.
+
+  The structure of this document is as follows:
+
+  - ``id``: Required string. Unique name for this entity. The YAML file SHOULD
+    define a `node anchor`_ for this field (e.g. ``id: &bucket0 bucket0``).
+
+  - ``database``: Required string. Database entity from which this bucket will
+    be created. The YAML file SHOULD use an `alias node`_ for a database
+    entity's ``id`` field (e.g. ``database: *database0``).
+
+  - ``bucketOptions``: Optional document. Additional options used to construct
+    the bucket object. Supported options are defined in the
+    `GridFS <../source/gridfs/gridfs-spec.rst#configurable-gridfsbucket-class>`__
+    specification. The ``readConcern``, ``readPreference``, and ``writeConcern``
+    options use the same structure as defined in `Common Options`_.
+
+
+collectionData
+~~~~~~~~~~~~~~
+
+List of documents that should correspond to the contents of a collection. This
+structure is used by both `initialData`_ and `test.outcome <test_outcome_>`_,
+which insert and read documents, respectively.
+
+The structure of this document is as follows:
+
+- ``collectionName``: Optional string. Collection name (not an `entity`_). The
+  YAML file SHOULD use an `alias node`_ for this value (e.g.
+  ``collectionName: *collection0Name``). Defaults to the name of the collection
+  under test (see: `collectionName`_).
+
+- ``databaseName``: Optional string. Database name (not an `entity`_). The
+  YAML file SHOULD use an `alias node`_ for this value (e.g.
+  ``databaseName: *database0Name``). Defaults to the name of the database under
+  test (see: `databaseName`_).
+
+- ``documents``: Required array of documents. List of documents corresponding to
```

Comment:
I did not plan to design that as it's not currently needed by any spec tests.

If it were to be implemented, I think it might be problematic for the `collectionData` struct, which is shared by `initialData` and `outcome`. The `documents` array here uses no special syntax and is a literal representation of what will be inserted or queried.

We could introduce an operator that takes a `value` string and `repeat` integer and evaluates to a repeated string. A variant of that could also be created for expressing or matching hex bytes. This operator could be used for matching (as with other operators), but I think it'd be problematic to introduce operators into arbitrary arguments (e.g. `document` or `documents` for `insertOne` and `insertMany`, respectively). Rather than attempt to shoe-horn this into regular operations, it may be preferable to define custom `operations` for collection entities with our own options -- similar to what we did for `createChangeStream`. That would allow more control over how to craft large bulk writes without requiring test runners to apply special processing to regular API methods.

For testing uploaded GridFS files, we do have the required `hexBytes` option for creating a `stream` entity. _That_ could be changed to also take a `repeat` option alongside the byte string. The `$$matchesHexBytes` operator could also be changed in a similar manner, so we could assert the contents of large files.

Frankly, I don't think this is needed for GridFS, since we can already test chunk-splitting with custom `chunkSize` options. Bulk writes are another story.

Reply:
Bulk writes are my major concern as there is much opportunity for implementations mistakes in splitting implementations and no way way to artificially reduce the batch sizes.

Reply:
Is the concern in being able to specify the required data at all, or in specifying it in a human-readable format?

If it's the former, doing something like gzip compressing an extended json representation of the data should make it fairly easy to work with, and universal, but the compressed bytes won't be human-readable.

Reply:
IMO, it'd be preferable to test bulk write batch splitting via prose tests. I expect many would disagree with that statement so I'll entertain a suggested approach.

There are several situations where a batch might be split (e.g. large documents in inserts, a multitude of small operations). Each of these would likely require unique syntax to express a large document (e.g. repeat a string X times) or multiple operations (e.g. repeat a sequence of WriteModels in a the bulk write's `requests` array). The spec currently has no instance of special rules to prepare arguments beyond `session`, `readConcern`, `readPreference`, and `writeConcern` -- all of which are considered "common options". If tests for bulk writes _were_ to be added to the spec, I expect it'd be implemented as more custom operations on the collection entity. Consider:

```
name: bulkWriteRepeatedRequests
object: collection0
arguments:
    requests: [...]
    repeatRequests: 1000
```

In any event, this seems like an additive change that would entirely self-contained within a new operation type. If we can agree to punt on this, I'll just capture this suggestion as a new Future Work section.

Reply:
I support punting until later.

On the question of whether splitting should be handled via prose or spec tests, I do not have an opinion at this time but I know that in Ruby we had a number of tickets (some user-reported) related to splitting and we ended up doing multiple passes of both changes and tests to ensure we do everything correctly, therefore this seems like an issue worthy of its own discussion/ticket/spec change.

Reply:
Another complication with splitting is the type of error produced - BSON size limit exceeded is a client error and generally driver-specific.

Reply:
I'm ok to punt to future work